### ISSUE TITLE
Styleflow hyperparameters

### ISSUE THREAD

**Issue by imthebilliejoe** [gitlabspy/styleflow#3]
Hi, I was able to run a version of styleflow in tf2 similar to yours, and I obtain good (but not good enough) results using 10k images of which i calculated (with external models) age and gender.

I train the model (styleflow) for 20 epoches with batch=64 but when I do forward and reverse (with the same attributes) the resulting w (and the image I obtain from it in Stylegan2) is slightly different while I was expecting to be exactly the same.

What do you think I'm still missing? 

Some ideas:
Should I probably go for more epoches? 
are only 2 attributes not enough? (on the paper they say that using only one attribute would overfit the model)
On the paper they go for 10 epoches and batch of 5(!!!) but with this setting the model seems to not converge (with my data).

**Comment by gitlabspy** [comment]
FFJORD is one special kind of flows. It does **pseudo inversion** ( not a very precise term) since the Jacobian is estimated. One workaround in StyleFlow is using style mixing, instead of generating an image with the resulting w, replace some of the layers of original w. 

For converge problem, do you use moving batchnorm? Batchnorm will speed up the converging (or any norm, or any affine transform...). 2 attributes is enough but disentanglement might not be good as more attributes.

**Comment by zanonShao** [comment]
wow

**Comment by imthebilliejoe** [comment]
Thank you for your answer. I'm doing what I think is the closest thing to what I read in the paper so yes, I'm using movingBatch before and after the conditional ffjord and use style mixing as specified in the paper (4 - 7 for age for example). Anyway my question came from the fact that when I use the ws that came up from styleflow (with same attribute) they look very far from the original Image. Some example:
original image:
![image](https://user-images.githubusercontent.com/80759886/112453451-fced1880-8d57-11eb-922b-c431e755c0c0.png)
image obtained using the whole new w from styleflow (no stylemixing) with same attributes:
![image](https://user-images.githubusercontent.com/80759886/112473742-0766dc80-8d6f-11eb-9356-6699f56543b1.png)
Image obtained with the age style mixing (4-7) but same attributes (this should be very similar to the input):
![image](https://user-images.githubusercontent.com/80759886/112453370-e8a91b80-8d57-11eb-9f7d-f8bccabec1bc.png)
image obtained changing the inclination of the face with style mixing (0-3):
![image](https://user-images.githubusercontent.com/80759886/112473702-fa49ed80-8d6e-11eb-8863-1da9240c542d.png)


Looking at the second image I was wondering if it's normal that it is so different from the input image or if I should obtain something better or more similar..

**Comment by zhanghongyong123456** [comment]
> 谢谢您的回答。我正在做的事情与我在本文中所读的内容最接近，所以是的，我在有条件的峡之前和之后使用movingBatch，并使用论文中指定的样式混合（例如，年龄为4-7） 。无论如何，我的问题来自这样一个事实，当我使用来自样式流（具有相同属性）的ws时，它们看起来与原始Image距离很远。例如：
> 原始图像： 使用来自newflow的全新w从样式流（无 样式混合）获得的，具有相同属性的图像：使用年龄样式混合（4-7）但具有相同的属性（与输入非常相似）获得的 图像：image通过样式混合（0-3）获得了改变面部倾斜度的结果：
> ![图像](https://user-images.githubusercontent.com/80759886/112453451-fced1880-8d57-11eb-922b-c431e755c0c0.png)
> 
> ![图像](https://user-images.githubusercontent.com/80759886/112473742-0766dc80-8d6f-11eb-9356-6699f56543b1.png)
> 
> ![图像](https://user-images.githubusercontent.com/80759886/112453370-e8a91b80-8d57-11eb-9f7d-f8bccabec1bc.png)
> 
> ![图像](https://user-images.githubusercontent.com/80759886/112473702-fa49ed80-8d6e-11eb-8863-1da9240c542d.png)
> 
> 看着第二张图片，我想知道它与输入图片有什么不同是正常的，还是我应该获得更好或更相似的东西。

I also want to try to train my own attributes, but I don’t know how to do it, can you give me some pointers on how to achieve it

**Comment by gitlabspy** [comment]
From your second image,  I assume your model generates an average face. I encounter such issues as well. It might probably be caused the batchnorm when you use a batchsize of 1 for inference. I don't know how to deal with this actually, but I found a  workaround that is using some other normalization (actnorm). Actually the affine transform in batchnorm is the only thing that speed up the converge, so you might need to write a new norm transform with only a affine transform (maybe with some init like actnorm does).

**Comment by gitlabspy** [comment]
Here's the code that I copy from tfp.glow, I replace batchnorm with this one and everything goes well. 
```
class ActivationNormalization(tfb.Bijector):
    def __init__(self, nchan, dtype=tf.float32, validate_args=False, name=None):
        parameters = dict(locals())
        self._initialized = tf.Variable(False, trainable=False)
        self._m = tf.Variable(tf.zeros(nchan, dtype))
        self._s = tfp.util.TransformedVariable(tf.ones(nchan, dtype), tfb.Exp())
        self._bijector = tfb.Invert(
            tfb.Chain([
                tfb.Scale(self._s),
                tfb.Shift(self._m),
            ]))
        super(ActivationNormalization, self).__init__(
            validate_args=validate_args,
            forward_min_event_ndims=1,
            parameters=parameters,
            name=name or 'ActivationNormalization')

    def _inverse(self, y,**kwargs):
        with tf.control_dependencies([self._maybe_init(y, inverse=True)]):
            return self._bijector.inverse(y, **kwargs)

    def _forward(self, x, **kwargs):
        with tf.control_dependencies([self._maybe_init(x, inverse=False)]):
            return self._bijector.forward(x, **kwargs)

    def _inverse_log_det_jacobian(self, y,  **kwargs):
        with tf.control_dependencies([self._maybe_init(y, inverse=True)]):
            return self._bijector.inverse_log_det_jacobian(y, 1, **kwargs)

    def _forward_log_det_jacobian(self, x, **kwargs):
        with tf.control_dependencies([self._maybe_init(x, inverse=False)]):
            return self._bijector.forward_log_det_jacobian(x, 1, **kwargs)

    def _maybe_init(self, inputs, inverse):
        def _init():
            axis = prefer_static.range(prefer_static.rank(inputs) - 1)
            m = tf.math.reduce_mean(inputs, axis=axis)
            s = (
                tf.math.reduce_std(inputs, axis=axis) +
                10. * np.finfo(dtype_util.as_numpy_dtype(inputs.dtype)).eps)
            if inverse:
                s = 1 / s
                m = -m
            else:
                m = m / s
            with tf.control_dependencies([self._m.assign(m), self._s.assign(s)]):
                return self._initialized.assign(True)

        return tf.cond(self._initialized, tf.no_op, _init)

```

**Comment by imthebilliejoe** [comment]
this worked incredibly well for me!!! thank you!

input image:
![image](https://user-images.githubusercontent.com/80759886/112612510-4bb1b580-8e1f-11eb-8335-bb8dffbbe6d5.png)
image obtained using the whole new w from styleflow (no stylemixing) with same attributes:
![image](https://user-images.githubusercontent.com/80759886/112612635-6e43ce80-8e1f-11eb-894c-87c0b363f22c.png)
Image obtained with the age style mixing (4-7) but same attributes (it's practical identical!):
![image](https://user-images.githubusercontent.com/80759886/112612697-81569e80-8e1f-11eb-9e41-20b70b946dee.png)

I dont know why in the paper they say to use batchnorm instead of actnorm if these are the results.
Really appreciate your help.

**Comment by gitlabspy** [comment]
😃  Great! I'll close it for now. If you encounter any issues, feel free to ask and I'll try my best to answer them.
